### PR TITLE
.*: update prometheus client, remove deprecated InstrumentHandler and InstrumentHandlerFunc  methods

### PR DESCRIPTION
**Pull request by kakkoyun**
Updates `prometheus/client_golang`. Remove usages of deprecated methods by introducing wrapper methods which mimics behaviour the same behaviour using new APIs.

## Changes

* Updates `prometheus/client_golang`.
* Removes usages of `prometheus.InstrumentHandler` and `prometheus. InstrumentHandlerFunc`
* Introduces `extprom/NewInstrumentedHandler` and `extprom/NewInstrumentedHandlerFunc`.

## Verification

By using `make test`, it seems like previous behaviour kept intact.

### CODE REVIEW

**Comment by metalmatze** [comment]
Looking really good. 
What was the intention to call the package `extprom` over `promhttp` or something to make it more obvious that the package is about http metrics?

**Comment by kakkoyun** [comment]
@metalmatze `extprom` was already there, which I believe to provide some extra behaviour to prometheus, it seemed like appropriate. I also have my doubts. I can move it to `promhttp` or to be more specific I can go with `promhttpserver`?

File: pkg/extprom/instrument_server.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+		prometheus.SummaryOpts{
+			Name: "http_response_size_bytes",
+			Help: "Tracks the size of HTTP responses.",
+		},
+		[]string{"code", "handler", "method"},
+	)
+)
+
+func init() {
+	prometheus.MustRegister(requestDuration, requestSize, requestsTotal, responseSize)
```
Comment (brancz) [inline pkg/extprom/instrument_server.go]: generally in thanos there are no init functions used for instrumentation purposes, they are instantiated and registered where used and not global (generally a good practice that we're also pushing everywhere else :) )

Best would be to create this at the very top and pass it down into each function/router that wants to instrument its handler.

**Comment by metalmatze** [comment]
> @metalmatze `extprom` was already there, which I believe to provide some extra behaviour to prometheus, it seemed like appropriate. I also have my doubts. I can move it to `promhttp` or to be more specific I can go with `promhttpserver`?

Just asking. I am totally fine with the name, if that's what you want. :wink:

File: pkg/extprom/instrument_server.go (hunk 1 above)
Comment (kakkoyun) [reply pkg/extprom/instrument_server.go]: I have removed `init` and package level vars and introduced new a new type to wrap metrics. I don't think `Instrumentor` is an actual word hence I'm open to suggestions. I used `Register` functions to inject this new behaviour, still not sure if this solution elegant enough 🤔

**Comment by brancz** [comment]
All of this lgtm, I think we can move forward and propose it upstream, I'm sure there will be some naming/package comments, but that's ok. Generally the strategy here is great! Nice job! :tada:

File: pkg/extprom/http/instrument_server.go [hunk 2]
```diff
@@ -72,7 +72,7 @@ func NewServerInstrumentor() ServerInstrumentor {
 			[]string{"code", "handler", "method"},
 		),
 	}
-	prometheus.MustRegister(ins.requestDuration, ins.requestSize, ins.requestsTotal, ins.responseSize)
+	reg.MustRegister(ins.requestDuration, ins.requestSize, ins.requestsTotal, ins.responseSize)
```
Comment (brancz) [inline pkg/extprom/http/instrument_server.go:75]: this looks like its gonna panic with multiple registrations

Comment (kakkoyun) [reply pkg/extprom/http/instrument_server.go:75]: That's what I've thought as well. I kept the existing behaviour from prometheus handler, I assumed if it worked, this should be fine. I will send the PR upstream and then I'll have another look at it.

**Pull request closed without merging**